components/registry/org.wso2.carbon.registry.extensions/src/main/java/org/wso2/carbon/registry/extensions/handlers/utils/WADLProcessor.java

```diff
@@ -117,38 +118,60 @@ public String addWadlToRegistry(RequestContext requestContext, Resource resource
 
         OMElement grammarsElement = wadlElement.
                 getFirstChildWithName(new QName(wadlNamespace, "grammars"));
-        if (grammarsElement != null) {
-            //This is to avoid evaluating the grammars import when building AST
-            grammarsElement.detach();
-        }
 
-        if(!skipValidation)  {
-            File tempFile = null;
-            BufferedWriter bufferedWriter = null;
-            try{
-                tempFile = File.createTempFile(wadlName, null);
-                bufferedWriter = new BufferedWriter(new FileWriter(tempFile));
-                bufferedWriter.write(wadlElement.toString());
-                bufferedWriter.flush();
-            } catch (IOException e) {
-                String msg = "Error occured while reading the WADL File";
-                log.error(msg);
-                throw new RegistryException(msg, e);
-            } finally {
-                if (bufferedWriter != null) {
-                    try {
-                        bufferedWriter.close();
-                    } catch (IOException e) {
-                        log.error("Error occurred while closing File writer");
+        if (StringUtils.isNotBlank(requestContext.getSourceURL())) {
+            String uri = requestContext.getSourceURL();
+            if (!skipValidation) {
+                validateWADL(uri);
+            }
+
+            if (resource.getUUID() == null) {
+                resource.setUUID(UUID.randomUUID().toString());
+            }
+
+            String wadlBaseUri = uri.substring(0, uri.lastIndexOf("/") + 1);
+            if (grammarsElement != null) {
+                //This is to avoid evaluating the grammars import when building AST
+                grammarsElement.detach();
+                wadlElement.addChild(resolveImports(grammarsElement, wadlBaseUri, version));
+            }
+        } else {
+            if (!skipValidation) {
+                File tempFile = null;
+                BufferedWriter bufferedWriter = null;
+                try {
+                    tempFile = File.createTempFile(wadlName, null);
+                    bufferedWriter = new BufferedWriter(new FileWriter(tempFile));
+                    bufferedWriter.write(wadlElement.toString());
+                    bufferedWriter.flush();
+                } catch (IOException e) {
+                    String msg = "Error occurred while reading the WADL File";
```

Comment:
Thanks for pointing out this. This is fixed in https://github.com/wso2/carbon-registry/pull/68